core/src/solution_submission.rs

```diff
@@ -174,77 +174,108 @@ impl<'a> StableXSolutionSubmitting for StableXSolutionSubmitter<'a> {
+            }
+            Err(ref err) => {
+                log::warn!(
+                    "failed to get gas price from gnosis safe gas station: {}",
+                    err
+                );
+            }
+        };
+        self.previous_gas_price
+    }
+}
+
+fn gas_price(estimated_price: U256, price_increase_count: u32, cap: U256) -> U256 {
+    let factor = 2u32.pow(price_increase_count);
+    cap.min(estimated_price * factor)
+}
+
 async fn retry_with_gas_price_increase(
     contract: &(dyn StableXContract + Sync),
     batch_index: u32,
     solution: Solution,
     claimed_objective_value: U256,
     gas_price_estimating: &(dyn GasPriceEstimating + Sync),
-    gas_cap: U256,
+    gas_price_cap: U256,
 ) -> Result<(), MethodError> {
-    const INCREASE_FACTOR: u32 = 2;
     const BLOCK_TIMEOUT: usize = 2;
     const DEFAULT_GAS_PRICE: u64 = 15_000_000_000;
-
     // openethereum requires that the gas price of the resubmitted transaction has increased by at
     // least 12.5%.
-    let min_gas_price_increase_factor: f64 = 1.125 * (1.0 + f64::EPSILON);
-    let effective_gas_cap =
-        U256::from((gas_cap.as_u128() as f64 / min_gas_price_increase_factor).floor() as u128);
-
-    let mut gas_price_estimate = U256::from(DEFAULT_GAS_PRICE);
-    let mut gas_price_factor = 1;
-    let mut result;
-    // the following block emulates a do-while loop
-    while {
-        gas_price_estimate = match gas_price_estimating.estimate_gas_price().await {
-            Ok(gas_estimate) => gas_estimate.fast,
-            Err(ref err) => {
-                log::warn!(
-                    "failed to get gas price from gnosis safe gas station: {}",
-                    err
-                );
-                gas_price_estimate
-            }
-        };
-        // Never exceed the gas cap.
-        let gas_price = std::cmp::min(gas_price_estimate * gas_price_factor, gas_cap);
+    const MIN_GAS_PRICE_INCREASE_FACTOR: f64 = 1.125 * (1.0 + f64::EPSILON);
 
-        // Submit solution at estimated gas price (not exceeding the cap)
-        result = contract
+    let effective_gas_price_cap = U256::from(
+        (gas_price_cap.as_u128() as f64 / MIN_GAS_PRICE_INCREASE_FACTOR).floor() as u128,
+    );
+    assert!(effective_gas_price_cap <= gas_price_cap);
+
+    let mut gas_price_estimator =
+        InfallibleGasPriceEstimator::new(gas_price_estimating, DEFAULT_GAS_PRICE.into());
+
+    for gas_price_increase_count in 0u32.. {
+        let estimated_price = gas_price_estimator.estimate().await;
+        let gas_price = gas_price(estimated_price, gas_price_increase_count, gas_price_cap);
+        assert!(gas_price <= gas_price_cap);
+        info!(
+            "solution submission try {} with gas price {}",
+            gas_price_increase_count, gas_price
+        );
+        let is_last_iteration = gas_price >= effective_gas_price_cap;
+        let block_timeout = if is_last_iteration {
+            None
```

Comment:
Now that I think of it, it may be a little dangerous to have `None` as a block timeout for confirmations. The reason being it can be that there is a gas price spike and it takes minutes (or even hours) for the transaction to be mined. Maybe we should use a more reasonable block timeout, something like 5 to 10 (75 - 150 seconds on average). In that case `is_last_iteration` would be needed.

On the other hand, we will be selecting this future with a timeout future, so it probably doesn't really matter.